[feat] Enable launchers to forward environment variables

jack-morrison opened the pull request:
Support for environment variable forwarding is added to the `srun` launcher, as well as two new launchers `mpirun-openmpi` and `mpirun-intelmpi`.

Closes #3207.

References: 
* [`srun --export`](https://slurm.schedmd.com/srun.html#OPT_export)
* [IntelMPI mpirun (hydra) options](https://www.intel.com/content/www/us/en/docs/mpi-library/developer-reference-linux/2021-15/global-hydra-options.html), see `-genv`
* [OpenMPI `mpirun -x`](https://docs.open-mpi.org/en/v5.0.x/man-openmpi/man1/mpirun.1.html#the-x-option)

vkarak commented on reframe/core/launchers/__init__.py:
```diff
@@ -36,6 +36,14 @@ class JobLauncher(metaclass=_JobLauncherMeta):
+    #:
+    #: :type: :class:`Dict[str, str]`
+    #: :default: ``{}``
+    environment_variables = variable(typ.Dict[str, str], value={})
```
To be consistent with the test's `env_vars`:

```suggestion
    env_vars = variable(typ.Dict[str, str], value={})
```

vkarak commented on reframe/core/launchers/__init__.py:
```diff
@@ -73,8 +82,6 @@ def command(self, job):
     def run_command(self, job):
         '''The full launcher command to be emitted for a specific job.
 
-        This includes any user options.
```
I think this is still accurate:

https://github.com/reframe-hpc/reframe/blob/cb4a2ac751c0f19637d735e4009dd20518d3a2af/reframe/core/launchers/__init__.py#L86

Maybe we should be more specific by writing "user-specific launcher options."

vkarak commented on reframe/core/launchers/mpi.py:
```diff
@@ -48,8 +48,11 @@ def command(self, job):
 
-        return ret
+        if self.environment_variables:
+            env_vars = ','.join(f'{k}={v}' for k, v in self.environment_variables.items())
```
This line is too long but it may fit once you rename `environment_variables` to `env_vars`.

vkarak commented on reframe/core/launchers/mpi.py:
```diff
@@ -108,6 +111,21 @@ class MpirunLauncher(JobLauncher):
+    def command(self, job):
+        cmd = ['mpirun', '-np', str(job.num_tasks)]
+        for name, value in self.environment_variables.items():
+            cmd += ['-x', f'{name}={value}']
```
Style issue: leave a blank line after this one.

vkarak commented on reframe/core/launchers/mpi.py:
```diff
@@ -108,6 +111,21 @@ class MpirunLauncher(JobLauncher):
+    def command(self, job):
+        cmd = ['mpirun', '-np', str(job.num_tasks)]
+        for name, value in self.environment_variables.items():
+            cmd += ['-genv', name, value]
```
Style issue: leave a blank line after this one.

vkarak commented on unittests/test_launchers.py:
```diff
@@ -127,7 +128,11 @@ def test_run_command(job):
         assert command == 'mpirun -np 4 --foo'
     elif launcher_name == 'srun':
-        assert command == 'srun --cpus-per-task=2 --foo'
+        assert command == 'srun --cpus-per-task=2 --export=FOO=bar,BAR=baz --foo'
```
Style issue: this line is too long.

vkarak commented on reframe/core/launchers/__init__.py:
```diff
@@ -36,6 +36,14 @@ class JobLauncher(metaclass=_JobLauncherMeta):
 
+    #: Dictionary of environment variables to be passed via the job launcher
+    #: invocation. The keys are the variable names and the values are their
+    #: corresponding values.
```
We should expand this mentioning which backends respect and properly interpret this variable.

vkarak reviewed (changes requested):
Mostly style comments. We would also need to update the general docs about launchers here:

https://github.com/reframe-hpc/reframe/blob/cb4a2ac751c0f19637d735e4009dd20518d3a2af/docs/config_reference.rst?plain=1#L495

Where we need to add the new launchers.

jack-morrison commented:
I believe I've addressed all review comments via [4166dcf](https://github.com/reframe-hpc/reframe/pull/3464/commits/4166dcf10df6500003d8da0d26d83790a10b0442), but still need to manually retest following this commit.

vkarak commented on reframe/core/launchers/__init__.py:
```diff
@@ -36,6 +36,20 @@ class JobLauncher(metaclass=_JobLauncherMeta):
+    #: - ``srun``
+    #: - ``mpirun-openmpi``
+    #: - ``mpirun-intelmpi``
+    #:
```
Add also a `.. versionadded:: 4.9` annotation here.

vkarak reviewed (changes requested):
We still need to add the new launchers in the configuration docs.

vkarak reviewed (approved):
I did the remaining small changes and rebased the PR. Lgtm now. Thanks!
